tests/Graph/TraverserCacheTest.cpp

```diff
@@ -0,0 +1,205 @@
+  std::shared_ptr<transaction::Context> queryContext{nullptr};
+  std::unique_ptr<arangodb::transaction::Methods> trx{nullptr};
+  arangodb::ResourceMonitor* _monitor;
+
+  TraverserCacheTest() : gdb(s.server, "testVocbase") {
+    query = gdb.getQuery("RETURN 1", std::vector<std::string>{});
+    queryContext = query.get()->newTrxContext();
+    trx = std::make_unique<arangodb::transaction::Methods>(queryContext);
+    _monitor = &query->resourceMonitor();
+    traverserCache =
+        std::make_unique<RefactoredTraverserCache>(trx.get(), query.get(),
+                                                   query->resourceMonitor(), stats);
+  }
+
+  ~TraverserCacheTest() = default;
+};
+
+TEST_F(TraverserCacheTest, it_should_return_a_null_aqlvalue_if_vertex_is_not_available) {
+  // prepare graph data - in this case, no data (no vertices and no edges, but collections v and e)
+  graph::MockGraph graph{};
+  gdb.addGraph(graph);
+
+  std::string vertexId = "v/Vertex";
+  std::string expectedMessage = "vertex '" + vertexId + "' not found";
+
+  auto data = VPackParser::fromJson(R"({"_key":"Vertex", "_id": "v/Vertex"})");
+  VPackSlice doc = data->slice();
+  HashedStringRef id{doc.get("_id")};
+  VPackBuilder builder;
+
+  // NOTE: we do not have the data, so we get null for any vertex
+  traverserCache->insertVertexIntoResult(arangodb::velocypack::HashedStringRef(id), builder);
+  ASSERT_TRUE(builder.slice().isNull());
+  auto all = query->warnings().all();
+  ASSERT_EQ(all.size(), 1);
+  ASSERT_TRUE(all[0].first == TRI_ERROR_ARANGO_DOCUMENT_NOT_FOUND);
+  ASSERT_TRUE(all[0].second == expectedMessage);
+}
+
+TEST_F(TraverserCacheTest, it_should_return_a_null_aqlvalue_if_edge_is_not_available) {
+  // prepare graph data - in this case, no data (no vertices and no edges, but collections v and e)
+  graph::MockGraph graph{};
+  gdb.addGraph(graph);
+
+  std::shared_ptr<arangodb::LogicalCollection> col =
+      gdb.vocbase.lookupCollection("e");
+  LocalDocumentId localDocumentId{1};    // invalid
+  DataSourceId dataSourceId{col->id()};  // valid
+  EdgeDocumentToken edt{dataSourceId, localDocumentId};
+  VPackBuilder builder;
+
+  // NOTE: we do not have the data, so we get null for any vertex
+  traverserCache->insertEdgeIntoResult(edt, builder);
+  ASSERT_TRUE(builder.slice().isNull());
+}
+
+TEST_F(TraverserCacheTest, it_should_increase_memory_usage_when_persisting_a_string) {
+  auto memoryUsageBefore = _monitor->current();
+
+  auto data = VPackParser::fromJson(R"({"_key":"123", "value":123})");
+  VPackSlice doc = data->slice();
+  HashedStringRef key{doc.get("_key")};
+  traverserCache->persistString(key);
+
+  EXPECT_LT(memoryUsageBefore, _monitor->current());
+  // now clear the class, and check memory usage again
+  traverserCache->clear();
+  EXPECT_EQ(memoryUsageBefore, _monitor->current());
+}
+
+TEST_F(TraverserCacheTest, it_should_not_increase_memory_usage_when_persisting_duplicate_string) {
```

Comment:
Please add a test that persists two strings.
Both should increase the memory.

Reply:
Please also add a test that:
Persistes one string. (Resources: x)
clears() (Resources: 0)
Then persists the same string again (Resources: x)

To avoid that we do not recount for the same string after a clear (e.g. forgot to erase lookup table)